### DISCUSSION TITLE
Return a value from a method

### DISCUSSION THREAD

**Discussion by KCD3V** [vfsfitvnm/frida-il2cpp-bridge#419] [Q&A] [answered]
`System.Int32 get_Width(); // 0x03971ed8`

```
[Token(Token = "0x6000306")]
[Address(RVA = "0x3971ED8", Offset = "0x3971ED8", VA = "0x3971ED8")]
public int get_Width()
{
	return 0;
}
```
When I do:
```
const a = cls.method("get_Width");
console.log('VAL', a);
```
It returns:
System.Int32 get_Width(); // 0x03971ed8

I've tried to invoke it, but it returns:
il2cpp: cannot invoke non-static method get_MaxWidth as it must be invoked throught a Il2Cpp.Object, not a Il2Cpp.Class.

What am I doing wrong?

**Comment by DragonRiseo** [comment]
what you did wrong was your brain
you did not think 
you want to return the method itself or the value of the method! 
then your return value . 
.....getwidth.... .value(); 
it would return 
1234

secondly since you hook it directly it wont work if it is non static , you need to implement it using anything even ```:any``` would work , does not have to be   Il2Cpp.Object, or  Il2Cpp.Class 

check the source code you would find more codes , have fun.

**Reply by KCD3V** [reply]
Ouch! Okay, I'll try some of your suggestions shortly and let you know if I manage to get a return value.

**Comment by KCD3V** [comment]
@DragonRiseo - please let me know what I'm doing wrong.

```
[Token(Token = "0x170000F4")]
public int MaxWidth
{
	[Token(Token = "0x60004BC")]
	[Address(RVA = "0x2465500", Offset = "0x2465500", VA = "0x2465500")]
	get
	{
		return 0;
	}
	[Token(Token = "0x60004BD")]
	[Address(RVA = "0x2465508", Offset = "0x2465508", VA = "0x2465508")]
	set
	{
	}
}
```

```
[Token(Token = "0x60004BC")]
[Address(RVA = "0x2465500", Offset = "0x2465500", VA = "0x2465500")]
public int get_MaxWidth()
{
	return 0;
}
```


Here are all my attempts at returning a value from that method so far:

Attempt 1: Returns nothing
```
ExampleCls.method("get_MaxWidth").implementation = function() {
        console.log('get_MaxWidth', this.method("get_MaxWidth").invoke());
        return this.method("get_MaxWidth").invoke();
}
```

Attempt 2: Returns the method, but no value: e.g. System.Int32 get_MaxWidth(); // 0x02465500
`const MaxWidth:any = ExampleCls.method("get_MaxWidth");`

Attempt 3: Returns undefined
`const MaxWidth:any = ExampleCls.method("get_MaxWidth").value;`

Attempt 4: Returns a TypeError: not a function
`const MaxWidth:any = ExampleCls.method("get_MaxWidth").value();`

Attempt 5: Returns "cannot invoke non-static method get_MaxWidth as it must be invoked throught a Il2Cpp.Object, not a Il2Cpp.Class"
`const MaxWidth:any = ExampleCls.method("get_MaxWidth").invoke();`

Attempt 6: Nothing returned to console
```
const getMaxWidth = ExampleCls.method("get_MaxWidth");
console.log('Address', getMaxWidth.virtualAddress); // <-- Returns 0x7ad92b2500
const listener = Interceptor.attach(getMaxWidth.virtualAddress, {
        onEnter: function(args) {
            console.log('onEnter');
        },
        onLeave: function(args) {
            console.log('onLeave');
        }
});
```
Attempt 7
```
const MaxWidthOffset = 0x2465500;
const getMaxWidthMethod = ExampleCls.method("get_MaxWidth");
const ptrValue = ptr(getMaxWidthMethod.virtualAddress).add(MaxWidthOffset);
console.log('Value', Memory.readInt(ptrValue)); // <-- Returns -1441397792 (Wrong)

```

**Comment by KCD3V** [comment]
Another thing, I find confusing is, if I do something like:
const maxWidthValue = this.field("_maxWidth").value;

I get:
il2cpp: cannot access instance field ExampleClass::_maxWidth from a class, use an object instead.

What does that mean?

**Comment by thinhbuzz** [comment]
You need to find an instance of the ExampleCls class

**Reply by KCD3V** [reply]
Mm... but "this" refers to the instances of the class right.

I'm able to return the value of this public DefaultField.
public const int DefaultField = 8;

But not.
private int _maxWidth;

I assume because it has no value maybe.

**Reply by thinhbuzz** [reply, marked as answer]
reference to the class instance when get_MaxWidth is called
but your problem is that you want to call get_MaxWidth, you need to find an instance of the class before you can call the function: instance.method('get_MaxWidth').invole()

**Reply by thinhbuzz** [reply]
https://github.com/vfsfitvnm/frida-il2cpp-bridge/wiki/Snippets#heap-scanning
one of the ways to find instances, note that you should not use this method in most cases, but here you should use it to understand how to use it.

**Reply by KCD3V** [reply]
Thank you thinhbuzz, but I think I'll have to abandon this. 
I tried both heap scan methods in that wiki before, but nothing was returned for the class I'm interested in. get_MaxWidth is only called within one class, so no idea how I'll ever invoke it to find what the return value is.

**Comment by DragonRiseo** [comment]
This is soooooo shit
To fix it is easy, 
Will type you a few stuff here

**Comment by DragonRiseo** [comment]
First read the fkn source , give it exactly 30 minutes reading , and learn 

- when you hook a method like 
Public int get_shit() 
Return 0 
Then you hook it without invoke , 
You just return the method.value , that is it , just like that 
// define assembly c sharp 
// define class 
// define method ( all codes in snippets ) of this source , and also of course in the source
Hook method directly // gives you its offset for whatever reason you might think of needing it ( you wont ) 
Hook method but add .value at the end 🔚 you get its value like 1234 

When hooking method like this 
Public void set_shit ( int someShit) 
Then you invoke it , inside ``:any`` , you can use il2cpp object or .class. But any covers it all

Normal thing like the get method above 
Assembly c sharp
Class 
Then method and define paramters 
(Param1:any) // any here too , cause the parameters may be int , bool, string or even literally shit , it would work once for all no need to change 
Then here for the void method you invoke 
M.... invoke(here goes the paramter value )
Like .....invoke(99999) ; 
The code also in snippets

For field here is the trick, like you mod it in lgl team with c++ it needs update value
So here too you update the value ( using ipdate or anything that is called , or even you get a function and call it and use it to mod the field when called , 
Or even create a new whole instance , trick the class that has the method , and refer to a whole new class and use anything there 
Like in fwd assault 
We have a class enum of weapons that has all of it fields like 
AK-47;
Famas; 
And nothing is there, all you do is just. create a new instance , trick the class of fields, and use it to call the player class , where there is a method called ( shoot ) 
So when you shoot it updates ( like an update  , but even better cause you control its amount , instead update is 60 times per second would cause huge lag ) 
And then call both the fields the weapons like famas 
And the function of drop weapon 
So evertime you click shoot 
You shoot a weapon out of your own instead of the bullet like you shoot famas , or ak or whatever instead bullets 
And if you hook bomb or grenade , you shoot grenades
For fields you just do that all in-capsulated  in the hooking of the update 
To hook update you just hook it like get method
And inside you dont get value, you call the value itslef , 
And inside the call you just pass the field value 
Like you did with the static one (default field =8 ) 
But since it is non static needs update , and hence you in-capsulated it already inside the update it works like a bullet 🚅

I am now on phone , i would have just shared my codes , but i am gonna sleep , 💤 , so later maybe i send  the codes of that here 
You can just create it if you learn the snippets and study the source code, you can just so anything to the game , 

Start with easy one 
Frontier wars 
Make there hp , and spawn player more than once , and create a simple 3d esp 
Then just move on to fwd assault, create whatever in your whole brain there 
But first hook ondetection of anticheat of codestage and return it null so it would not ban you

**Reply by KCD3V** [reply]
Ha thanks very much for all this information. It was really helpful! Goodness me.
Have a good sleep.

**Comment by KCD3V** [comment]
So basically, to invoke a non-static method from a class, I could do:
```
const newObj = Class.new();
console.log('MaxWidth', newObj.method('get_MaxWidth').invoke());
```
But what thinhbuzz suggested was actually more helpful in that I need to find an instance of the class in the game to get anything meaningful from the get_MaxWidth method.
